keps/sig-api-machinery/20190206-watch-bookmark.md

```diff
@@ -0,0 +1,184 @@
+watcher).
+
+The reason for that is the fact that even if the last item received by watcher
+has resourceVersion rv1, we may already know that there aren't any changes
+a given watcher is interested in up to rv2 (rv1 < rv2), but we don't have any
+way of communicating it to the watcher. As a result, when restarting a watch,
+client again sends rv1 as a starting point, and we process all events with
+resourceVersion between rv1 and rv2 again unnecessarily.
+
+The proposal presents a proper solution for that problem.
+
+### Goals
+
+- Reduce load on apiserver by minimizing amount of unnecessary watch events
+that need to be processed after restarting a watch.
+- Reduce amount of undesired "resource version too old" errors on reestablishing
+a watch.
+
+### Non-Goals
+
+The following are nice-to-haves, but not primary goals:
+
+- Improve overall watch throughput and/or latency.
+
+## Proposal
+
+We propose introducing a new type of watch event called `Bookmark`. With that
+change, the possible watch event types will be:
+```
+  Added    EventType = "ADDED"
+  Modified EventType = "MODIFIED"
+  Deleted  EventType = "DELETED"
+  Error    EventType = "ERROR"
+  Bookmark EventType = "BOOKMARK"
+```
+
+Watch event with type Bookmark will represent information that all the objects
+up to a given resourceVersion has been processed for a given watcher. So even
+if the last event of other types contained object with resourceVersion rv1,
+receiving a bookmark with resourceVersion rv2 means that there aren't
+any interesting objects for that watcher in between.
+
+Given that we don't want to wait for v2 version of watch API with that change,
+an obvious requirement for introducing it is backward compatibility. Currently
+watch Event type looks as following:
+```
+  type Event struct {
+    Type EventType
+    Object runtime.Object
+  }
+```
+
+As a result, we will represent bookmark event by setting Bookmark type and
+Object of appropriate type with just ObjectMeta.ResourceVersion field set.
+
+Unfortunately, such a change would break existing clients, as an example
+[official decoder][]. As a result, we will extend `ListOptions` (which is
+how we pass options to watch) with a boolean field where user can opt-in
+for watch bookmarks:
+```
+  type ListOptions struct {
+    ...
+    AllowBookmarks bool
+  }
+```
+We consciously make it just a boolean flag - this gives kube-apiserver an
+ability to choose when they should be send without setting any expectations
+on user side how frequently and when it would be happening. In particular
+client isn't guaranteed to get any bookmarks.
+Such addition to `ListOptions` is also safe from backward-compatibility
+point of view, old kube-apiserver will simply drop this field if set.
+
+Once the API is extended, we add a support for sending bookmarks to watchcache.
+The exact policy of sending them is to be determined, the ideas include:
+- send a bookmark if there weren't any event send to user in the last X
```

Comment:
maybe adding this one:

```
- send a bookmark if there weren't any event matched object for past N object processed.
```

Number of object processed is represent the optimization here more than time I think.